File: packages/alp-cli/src/cli/services/envelope.ts
// SPDX-License-Identifier: Apache-2.0

import {
  CLI_EXIT_CODE,
  CliEnvelope,
  CliExecutionResult,
  CliFormat,
  CliIssue,
} from "../models";

export function createFailureResult<TData>(
  command: string,
  format: CliFormat,
  exitCode: number,
  textLines: string[],
  issues: CliIssue[],
  data: TData,
): CliExecutionResult<TData> {
  return {
    format,
    exitCode,
    textLines: format === "json" ? [] : textLines,
    envelope: createEnvelope(
      command,
      {
        root: null,
        boardYaml: null,
      },
      data,
      issues,
      exitCode,
    ),
  };
}

export function createEnvelope<TData>(
  command: string,
  project: { root: string | null; boardYaml: string | null },
  data: TData,
  issues: readonly CliIssue[],
  exitCode: number,
): CliEnvelope<TData> {
  return {
    command,
    ok: exitCode === CLI_EXIT_CODE.success,
    exitCode,
    project,
    data,
    issues: [...issues],
  };
}

File: cli-rs/crates/alp-cli/src/envelope.rs
// SPDX-License-Identifier: Apache-2.0
//! Machine-readable result envelope — byte-compatible with the TypeScript
//! CLI's `CliEnvelope`. JSON mode writes exactly one of these to stdout.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Issue {
    pub code: String,
    pub severity: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Project {
    pub root: Option<String>,
    #[serde(rename = "boardYaml")]
    pub board_yaml: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Envelope<T: Serialize> {
    pub command: String,
    pub ok: bool,
    #[serde(rename = "exitCode")]
    pub exit_code: i32,
    pub project: Project,
    pub data: T,
    pub issues: Vec<Issue>,
}

impl<T: Serialize> Envelope<T> {
    pub fn new(
        command: &str,
        project: Project,
        data: T,
        issues: Vec<Issue>,
        exit_code: i32,
    ) -> Self {
        Self {
            command: command.to_string(),
            ok: exit_code == 0,
            exit_code,
            project,
            data,
            issues,
        }
    }

    /// Serialize to a single-line JSON document (stdout contract).
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("envelope serialization is infallible")
    }
}
